AssertionError when filtering columns of a MatrixTable

To report a bug, fill in the information below. 
For support and feature requests, please use the discussion forum: http://discuss.hail.is/

-------------------------------------------------------------------------------------------

### Hail version:
0.2

### What you did:
```
mt = hl.read_matrix_table('gs://...')
mt = mt.filter_cols(mt.used_in_pca_calculation)
print(mt.count_cols())
```
The MatrixTable has a lot of row and column annotations. `mt.used_in_pca_calculation` is a Boolean column annotation.

### What went wrong (all error messages here, including the full java stack trace):
```
Traceback (most recent call last):
  File "/tmp/8d5cc778-fdc7-4210-a60b-5efd1f67c45f/subset_genotype_pca.py", line 8, in <module>
    print(mt.count_cols())
  File "/home/hail/hail.zip/hail/matrixtable.py", line 1950, in count_cols
  File "/usr/lib/spark/python/lib/py4j-0.10.4-src.zip/py4j/java_gateway.py", line 1133, in __call__
  File "/home/hail/hail.zip/hail/utils/java.py", line 206, in deco
hail.utils.java.FatalError: AssertionError: assertion failed

Java stack trace:
java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:156)
	at is.hail.expr.MatrixValue.filterSamplesKeep(Relational.scala:110)
	at is.hail.expr.MatrixValue.filterCols(Relational.scala:133)
	at is.hail.expr.FilterCols.execute(Relational.scala:333)
	at is.hail.variant.MatrixTable.value$lzycompute(MatrixTable.scala:536)
	at is.hail.variant.MatrixTable.value(MatrixTable.scala:534)
	at is.hail.variant.MatrixTable.x$16$lzycompute(MatrixTable.scala:541)
	at is.hail.variant.MatrixTable.x$16(MatrixTable.scala:541)
	at is.hail.variant.MatrixTable.colValues$lzycompute(MatrixTable.scala:541)
	at is.hail.variant.MatrixTable.colValues(MatrixTable.scala:541)
	at is.hail.variant.MatrixTable.numCols(MatrixTable.scala:2378)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:280)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:214)
	at java.lang.Thread.run(Thread.java:748)

Hail version: devel-3cf3108
Error summary: AssertionError: assertion failed
```

--- a/python/hail/tests/test_api.py
+++ b/python/hail/tests/test_api.py
@@ -856,10 +856,10 @@ class MatrixTests(unittest.TestCase):
         self.assertTrue(kept.all(hl.is_defined(kept.x2) & (kept.x2 == kept.x * 10)))
         self.assertTrue(removed.all(hl.is_missing(removed.x2)))
 
-    def test_required_entries(self):
+    def test_filter_cols_required_entries(self):
         mt1 = hl.utils.range_matrix_table(10, 10, n_partitions=4)
         mt1 = mt1.filter_cols(mt1.col_idx < 3)
-        mt1.entries().show()
+        self.assertEqual(len(mt1.entries().collect()), 30)
 
     def test_vcf_regression(self):
         ds = hl.import_vcf(resource('33alleles.vcf'))

--- a/python/hail/tests/test_api.py
+++ b/python/hail/tests/test_api.py
@@ -856,6 +856,11 @@ class MatrixTests(unittest.TestCase):
         self.assertTrue(kept.all(hl.is_defined(kept.x2) & (kept.x2 == kept.x * 10)))
         self.assertTrue(removed.all(hl.is_missing(removed.x2)))
 
+    def test_required_entries(self):
+        mt1 = hl.utils.range_matrix_table(10, 10, n_partitions=4)
+        mt1 = mt1.filter_cols(mt1.col_idx < 3)
+        mt1.entries().show()
+
     def test_vcf_regression(self):
         ds = hl.import_vcf(resource('33alleles.vcf'))
         self.assertEqual(

--- a/src/main/scala/is/hail/expr/Relational.scala
+++ b/src/main/scala/is/hail/expr/Relational.scala
@@ -160,7 +160,7 @@ object MatrixIR {
     })
   }
 
-  def chooseColsWithArray(typ: MatrixType): (Type, (MatrixValue, Array[Int]) => MatrixValue) = {
+  def chooseColsWithArray(typ: MatrixType): (MatrixType, (MatrixValue, Array[Int]) => MatrixValue) = {
     val rowType = typ.rvRowType
     val keepType = TArray(+TInt32())
     val (rTyp, makeF) = ir.Compile[Long, Long, Long]("row", rowType,
@@ -171,10 +171,12 @@ object MatrixIR {
             ir.Ref("i")))))))
     assert(rTyp.isOfType(rowType))
 
+    val newMatrixType = typ.copy(rvRowType = coerce[TStruct](rTyp))
+
     val keepF = { (mv: MatrixValue, keep: Array[Int]) =>
       val keepBc = mv.sparkContext.broadcast(keep)
       mv.copy(colValues = keep.map(mv.colValues),
-        rvd = mv.rvd.mapPartitionsPreservesPartitioning(typ.orvdType) { it =>
+        rvd = mv.rvd.mapPartitionsPreservesPartitioning(newMatrixType.orvdType) { it =>
           val f = makeF()
           val keep = keepBc.value
           var rv2 = RegionValue()
@@ -187,10 +189,10 @@ object MatrixIR {
           }
         })
     }
-    (rTyp, keepF)
+    (newMatrixType, keepF)
   }
 
-  def filterCols(typ: MatrixType): (Type, (MatrixValue, (Annotation, Int) => Boolean) => MatrixValue) = {
+  def filterCols(typ: MatrixType): (MatrixType, (MatrixValue, (Annotation, Int) => Boolean) => MatrixValue) = {
     val (t, keepF) = chooseColsWithArray(typ)
     (t, {(mv: MatrixValue, p :(Annotation, Int) => Boolean) =>
       val keep = (0 until mv.nCols)
@@ -351,9 +353,7 @@ case class FilterCols(
     FilterCols(newChildren(0).asInstanceOf[MatrixIR], pred)
   }
 
-  val (rvtyp, filterF) = MatrixIR.filterCols(child.typ)
-
-  def typ: MatrixType = child.typ.copy(rvRowType=coerce[TStruct](rvtyp))
+  val (typ, filterF) = MatrixIR.filterCols(child.typ)
 
   def execute(hc: HailContext): MatrixValue = {
     val prev = child.execute(hc)
@@ -386,9 +386,7 @@ case class FilterColsIR(
     FilterColsIR(newChildren(0).asInstanceOf[MatrixIR], newChildren(1).asInstanceOf[IR])
   }
 
-  val (rvtyp, filterF) = MatrixIR.filterCols(child.typ)
-
-  def typ: MatrixType = child.typ.copy(rvRowType=coerce[TStruct](rvtyp))
+  val (typ, filterF) = MatrixIR.filterCols(child.typ)
 
   def execute(hc: HailContext): MatrixValue = {
     val prev = child.execute(hc)

--- a/src/main/scala/is/hail/expr/Relational.scala
+++ b/src/main/scala/is/hail/expr/Relational.scala
@@ -99,41 +99,6 @@ case class MatrixValue(
   }
 
   lazy val colValuesBc: Broadcast[IndexedSeq[Annotation]] = sparkContext.broadcast(colValues)
-
-  def chooseColsWithArray(keep: Array[Int]): MatrixValue = {
-    val rowType = typ.rvRowType
-    val keepType = TArray(+TInt32())
-    val (rTyp, makeF) = ir.Compile[Long, Long, Long]("row", rowType,
-      "keep", keepType,
-      body = InsertFields(ir.Ref("row"), Seq((MatrixType.entriesIdentifier,
-        ir.ArrayMap(ir.Ref("keep"), "i",
-          ir.ArrayRef(ir.GetField(ir.In(0, rowType), MatrixType.entriesIdentifier),
-            ir.Ref("i")))))))
-    assert(rTyp == rowType)
-
-    val keepBc = sparkContext.broadcast(keep)
-    copy(colValues = keep.map(colValues),
-      rvd = rvd.mapPartitionsPreservesPartitioning(typ.orvdType) { it =>
-        val f = makeF()
-        val keep = keepBc.value
-        var rv2 = RegionValue()
-
-        it.map { rv =>
-          val region = rv.region
-          rv2.set(region,
-            f(region, rv.offset, false, region.appendArrayInt(keep), false))
-          rv2
-        }
-      })
-  }
-
-  def filterCols(p: (Annotation, Int) => Boolean): MatrixValue = {
-    val keep = (0 until nCols)
-      .view
-      .filter { i => p(colValues(i), i) }
-      .toArray
-    chooseColsWithArray(keep)
-  }
 }
 
 object MatrixIR {
@@ -192,6 +157,47 @@ object MatrixIR {
         FilterColsIR(m,
           ApplyBinaryPrimOp(DoubleAmpersand(), pred1, pred2))
        */
+    })
+  }
+
+  def chooseColsWithArray(typ: MatrixType): (Type, (MatrixValue, Array[Int]) => MatrixValue) = {
+    val rowType = typ.rvRowType
+    val keepType = TArray(+TInt32())
+    val (rTyp, makeF) = ir.Compile[Long, Long, Long]("row", rowType,
+      "keep", keepType,
+      body = InsertFields(ir.Ref("row"), Seq((MatrixType.entriesIdentifier,
+        ir.ArrayMap(ir.Ref("keep"), "i",
+          ir.ArrayRef(ir.GetField(ir.In(0, rowType), MatrixType.entriesIdentifier),
+            ir.Ref("i")))))))
+    assert(rTyp.isOfType(rowType))
+
+    val keepF = { (mv: MatrixValue, keep: Array[Int]) =>
+      val keepBc = mv.sparkContext.broadcast(keep)
+      mv.copy(colValues = keep.map(mv.colValues),
+        rvd = mv.rvd.mapPartitionsPreservesPartitioning(typ.orvdType) { it =>
+          val f = makeF()
+          val keep = keepBc.value
+          var rv2 = RegionValue()
+
+          it.map { rv =>
+            val region = rv.region
+            rv2.set(region,
+              f(region, rv.offset, false, region.appendArrayInt(keep), false))
+            rv2
+          }
+        })
+    }
+    (rTyp, keepF)
+  }
+
+  def filterCols(typ: MatrixType): (Type, (MatrixValue, (Annotation, Int) => Boolean) => MatrixValue) = {
+    val (t, keepF) = chooseColsWithArray(typ)
+    (t, {(mv: MatrixValue, p :(Annotation, Int) => Boolean) =>
+      val keep = (0 until mv.nCols)
+        .view
+        .filter { i => p(mv.colValues(i), i) }
+        .toArray
+      keepF(mv, keep)
     })
   }
 }
@@ -345,7 +351,9 @@ case class FilterCols(
     FilterCols(newChildren(0).asInstanceOf[MatrixIR], pred)
   }
 
-  def typ: MatrixType = child.typ
+  val (rvtyp, filterF) = MatrixIR.filterCols(child.typ)
+
+  def typ: MatrixType = child.typ.copy(rvRowType=coerce[TStruct](rvtyp))
 
   def execute(hc: HailContext): MatrixValue = {
     val prev = child.execute(hc)
@@ -363,7 +371,7 @@ case class FilterCols(
       ec.setAll(localGlobals, sa)
       f() == true
     }
-    prev.filterCols(p)
+    filterF(prev, p)
   }
 }
 
@@ -378,7 +386,9 @@ case class FilterColsIR(
     FilterColsIR(newChildren(0).asInstanceOf[MatrixIR], newChildren(1).asInstanceOf[IR])
   }
 
-  def typ: MatrixType = child.typ
+  val (rvtyp, filterF) = MatrixIR.filterCols(child.typ)
+
+  def typ: MatrixType = child.typ.copy(rvRowType=coerce[TStruct](rvtyp))
 
   def execute(hc: HailContext): MatrixValue = {
     val prev = child.execute(hc)
@@ -410,7 +420,7 @@ case class FilterColsIR(
       val colRVoffset = colRVb.end()
       predCompiledFunc()(colRegion, globalRVoffset, false, colRVoffset, false)
     }
-    prev.filterCols(p)
+    filterF(prev, p)
   }
 }
 

--- a/src/main/scala/is/hail/expr/Relational.scala
+++ b/src/main/scala/is/hail/expr/Relational.scala
@@ -175,7 +175,8 @@ object MatrixIR {
 
     val keepF = { (mv: MatrixValue, keep: Array[Int]) =>
       val keepBc = mv.sparkContext.broadcast(keep)
-      mv.copy(colValues = keep.map(mv.colValues),
+      mv.copy(typ = newMatrixType,
+        colValues = keep.map(mv.colValues),
         rvd = mv.rvd.mapPartitionsPreservesPartitioning(newMatrixType.orvdType) { it =>
           val f = makeF()
           val keep = keepBc.value
@@ -523,12 +524,12 @@ case class ChooseCols(child: MatrixIR, oldIndices: Array[Int]) extends MatrixIR 
     ChooseCols(newChildren(0).asInstanceOf[MatrixIR], oldIndices)
   }
 
-  def typ: MatrixType = child.typ
+  val (typ, colsF) = MatrixIR.chooseColsWithArray(child.typ)
 
   def execute(hc: HailContext): MatrixValue = {
     val prev = child.execute(hc)
 
-    prev.chooseColsWithArray(oldIndices)
+    colsF(prev, oldIndices)
   }
 }
 

--- a/src/main/scala/is/hail/variant/MatrixTable.scala
+++ b/src/main/scala/is/hail/variant/MatrixTable.scala
@@ -1343,7 +1343,8 @@ class MatrixTable(val hc: HailContext, val ast: MatrixIR) {
   }
 
   def filterCols(p: (Annotation, Int) => Boolean): MatrixTable = {
-    copyAST(ast = MatrixLiteral(matrixType, value.filterCols(p)))
+    val (newType, filterF) = MatrixIR.filterCols(matrixType)
+    copyAST(ast = MatrixLiteral(newType, filterF(value, p)))
   }
 
   def filterColsExpr(filterExpr: String, keep: Boolean = true): MatrixTable = {

